fms/modules/tp.py

```diff
@@ -31,14 +31,328 @@ def setup_tp(self, rank: int, world_size: int) -> None:
+            [0 4] [1 5] [2 6] [3 7]
+
+            world_size = 4, max_partition_sizes = [4, 1], tensor = [0 1 2 3 4 5 6 7 8 9]
+            [0 1 8 9] [2 3 8 9] [4 5 8 9] [6 7 8 9]
+
+        is_bias: bool
+            if True is bias, else is weight
+        """
+        # Divide the weight matrix along the first dimension.
+        cusum_max_partition_sizes = [0]
+        min_partition_size = min(max_partition_sizes)
+        for m in max_partition_sizes:
+            cusum_max_partition_sizes.append(
+                cusum_max_partition_sizes[-1] + (m // min_partition_size)
+            )
+
+        output_size_per_partition = param.shape[0] // (
+            sum(max_partition_sizes) // min(max_partition_sizes)
+        )
+        if not is_bias:
+            tensor = torch.cat(
+                [
+                    tensor_value[
+                        self.__get_start_index(
+                            tensor_value,
+                            output_size_per_partition,
+                            replication,
+                            max_partition_sizes,
+                            min_partition_size,
+                            cusum_max_partition_sizes,
+                            i,
+                        ) : self.__get_end_index(
+                            tensor_value,
+                            output_size_per_partition,
+                            replication,
+                            max_partition_sizes,
+                            min_partition_size,
+                            cusum_max_partition_sizes,
+                            i,
+                        ),
+                        :,
+                    ]
+                    for i, replication in enumerate(max_partition_sizes)
+                ],
+                dim=0,
+            )
+        else:
+            tensor = torch.cat(
+                [
+                    tensor_value[
+                        self.__get_start_index(
+                            tensor_value,
+                            output_size_per_partition,
+                            replication,
+                            max_partition_sizes,
+                            min_partition_size,
+                            cusum_max_partition_sizes,
+                            i,
+                        ) : self.__get_end_index(
+                            tensor_value,
+                            output_size_per_partition,
+                            replication,
+                            max_partition_sizes,
+                            min_partition_size,
+                            cusum_max_partition_sizes,
+                            i,
+                        )
+                    ]
+                    for i, replication in enumerate(max_partition_sizes)
+                ],
+                dim=0,
+            )
+        param.copy_(tensor, non_blocking=True)
+
+    def load(
```

Comment:
This is coming from what I know is going to happen for MoE, but I'd make the load function the one each module overrides instead of colwise_params and rowwise_params, and I'd have each module simply call `_copy_colwise` and the other functions for each of their specific modules. The reason for taht is that for the MoE modules, the whole matching string to parameters kinda breaks and I had to make some hacky changes, while it'd be a lot easier to simply be able to call the helper functions on the actual nn.Parameters directly.